sdc11073/mdib/devicemdib.py

```diff
@@ -495,56 +452,77 @@ def _process_transaction(self, setDeterminationTime):
+                    elif descriptorContainer.isComponentDescriptor:
+                        update_dict = mgr.componentStateUpdates
+                    elif descriptorContainer.isContextDescriptor:
+                        update_dict = mgr.contextStateUpdates
+                    elif descriptorContainer.isMetricDescriptor \
+                            and not descriptorContainer.isRealtimeSampleArrayMetricDescriptor:
+                        update_dict = mgr.metricStateUpdates
+                    elif descriptorContainer.isOperationalDescriptor:
+                        update_dict = mgr.operationalStateUpdates
+                    else:
+                        update_dict = mgr.rtSampleStateUpdates
+                    if descriptorContainer.isContextDescriptor:
+                        update_dict = mgr.contextStateUpdates
+                        all_states = self.contextStates.descriptorHandle.get(descriptorContainer.handle, [])
+                        for st in all_states:
+                            key = (descriptorContainer.handle, st.handle)
+                            # check if state is already present in this transaction
+                            state_update = update_dict.get(key)
+                            if state_update is not None:
+                                # the state has also been updated directly in transaction.
+                                # update descriptor version
+                                old_state, new_state = state_update
+                            else:
+                                old_state = st
+                                new_state = old_state.mkCopy()
+                                update_dict[key] = _TrItem(old_state, new_state)
+                            new_state.descriptorContainer = descriptorContainer
+                            new_state.incrementState()
+                            new_state.updateDescriptorVersion()
+                            descr_updated_states.append(new_state)
+                    else:
+                        # check if state is already present in this transaction
+                        state_update = update_dict.get(descriptorContainer.handle)
+                        new_state = None
+                        if state_update is not None:
+                            # the state has also been updated directly in transaction.
+                            # update descriptor version
+                            old_state, new_state = state_update
+                            if new_state is None:
+                                raise ValueError(f'state deleted? that should not be possible! handle = {descriptorContainer.handle}')
+                            new_state.descriptorContainer = descriptorContainer
+                            new_state.updateDescriptorVersion()
+                        else:
+                            old_state = self.states.descriptorHandle.getOne(descriptorContainer.handle, allowNone=True)

```

Comment:
This case should not be legal, a descriptor update without a state update is incomplete. Only in the case of contexts this actually indicates the state was deleted.

Reply:
This is the code that handles exactly that situation! It adds the corresponding state to the transaction.

Reply:
Yes, it should fail, not fix it ;)